### ISSUE TITLE
rare checkpointing test failure

### ISSUE THREAD

**Issue by benclifford** [Parsl/parsl#681]
I've seen this once with commit 1ac7adfa74f89e5dc46072ac03dd235633714d4a which is close to master commit a216c033dc30fe7b6a543f23b625c141a0ebc2f4.

It suggests that in some cases, checkpointing isn't writing out all values in time - perhaps in this situation, the last checkpoint wasn't flushed in time?

```
parsl/tests/test_checkpointing/test_regression_233.py ..F
=================================== FAILURES ===================================
_____________________ test_checkpoint_availability[local] ______________________
@pytest.mark.local
    @pytest.mark.forked
    def test_checkpoint_availability():
        import os
    
        original, run_dir = run_checkpointed([])
        last_checkpoint = os.path.join(run_dir, 'checkpoint')
        print(last_checkpoint)
        cached, _ = run_checkpointed([last_checkpoint])
    
        print(cached)
        print(original)
    
>       assert cached == original, "All tasks were not cached"
E       AssertionError: All tasks were not cached
E       assert [3077, 3508, 7971, 3621, 7635] == [3077, 3508, 7971, 3621, 9781]
E         At index 4 diff: 7635 != 9781
E         Use -v to get the full diff
```

**Comment by benclifford** [comment]
I just experienced this again in CI:

```
>       assert cached == original, "All tasks were not cached"
E       AssertionError: All tasks were not cached
E       assert [653, 3265, 4333, 3289, 8499] == [653, 3265, 4333, 3289, 2283]
E         At index 4 diff: 8499 != 2283
E         Use -v to get the full diff
```

**Comment by yadudoc** [comment]
I vote we defer this issue for 0.7.1, and skip the test for now.

**Comment by benclifford** [comment]
PR #696 disables this test.

This issue should remain open, because I think there is a likely bug in checkpointing shutdown here.

**Comment by benclifford** [comment]
tl;dr: this test makes incorrect assumptions about when checkpoints are written.

I'm able to force this test failure against current master (1ea85d84b3bab5e468c87eb5c4d63fd9eac780b7) by adding in two sleep statements (which should have no effect on behaviour other than slowing things down):

```
diff --git a/parsl/dataflow/dflow.py b/parsl/dataflow/dflow.py
index 9cdc48ecb..9c0bdadfb 100644
--- a/parsl/dataflow/dflow.py
+++ b/parsl/dataflow/dflow.py
@@ -331,6 +331,8 @@ class DataFlowKernel(object):
             self.memoizer.update_memo(task_id, self.tasks[task_id], future)
 
             if self.checkpoint_mode == 'task_exit':
+                import time
+                time.sleep(7)
                 self.checkpoint(tasks=[task_id])
 
         return
diff --git a/parsl/tests/test_checkpointing/test_regression_233.py b/parsl/tests/test_checkpointing/test_regression_233.py
index 9dc059d79..bc679c13a 100644
--- a/parsl/tests/test_checkpointing/test_regression_233.py
+++ b/parsl/tests/test_checkpointing/test_regression_233.py
@@ -13,6 +13,8 @@ def run_checkpointed(checkpoints):
     @App('python', dfk, cache=True)
     def cached_rand(x):
         import random
+        import time
+        time.sleep(1)
         return random.randint(0, 10000)
 
     items = []
```

The specific race condition that is happening, I think, is that a future returning from `future.result()` means that the result is available; it doesn't mean that other task completion stuff has happened. There is a related phenomenon with file stage-out: `future.result()` returning does not mean that file stage-out for that app has happened; merely that the actual result value is available and here it is.

Closing the DFK after results are available should cause all outstanding checkpoints to be written correclty; the offending test does not do this in the right order, though. It closes the DFK immediately after launching the apps, rather than after results are available.

By waiting for results before closing the DFK, this test works even in the presence of the above delays.

**Comment by benclifford** [comment]
This was fixed by #1250
